### PR TITLE
PATH Report

### PR DESCRIPTION
**Pull request by awisema**

### CODE REVIEW

File: app/models/concerns/arel_helper.rb [hunk 1]
```diff
@@ -82,6 +82,15 @@ def any(*args)
       nf 'ANY', args
     end
 
+    def array_agg(*args)
+      nf 'ARRAY_AGG', args
+    end
+
+    def array(*args)
```
Comment (eanders) [inline app/models/concerns/arel_helper.rb]: I wonder if this should be `sql_array` or something.  Seems like mixing in an `array` function wherever we include ArelHelper is a bit dangerous.

File: app/models/grda_warehouse/hud/current_living_situation.rb [hunk 2]
```diff
@@ -22,5 +22,10 @@ class CurrentLivingSituation < Base
     belongs_to :direct_client, **hud_assoc(:PersonalID, 'Client')
     belongs_to :data_source
 
+    scope :between, ->(start_date:, end_date:) do
+      at = arel_table
+      where(at[:InformationDate].gteq(start_date).and(at[:InformationDate].lteq(end_date)))
```
Comment (eanders) [inline app/models/grda_warehouse/hud/current_living_situation.rb]: Why not use `where(arel_table[:InformationDate].between(start_date, end_date))`

File: app/models/grda_warehouse/hud/service.rb [hunk 3]
```diff
@@ -42,6 +44,11 @@ class Service < Base
       )
     }
 
+    scope :between, ->(start_date:, end_date:) do
+      at = arel_table
+      where(at[:DateProvided].gteq(start_date).and(at[:DateProvided].lteq(end_date)))
```
Comment (eanders) [inline app/models/grda_warehouse/hud/service.rb]: same here with `where(arel_table[:DateProvided].between(start_date, end_date))`

File: drivers/hud_path_report/app/controllers/hud_path_report/cells_controller.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+module HudPathReport
+  class CellsController < HudPathReport::QuestionsController
+    before_action :set_report
+    before_action :set_question
+
+    def show
+      @cell = @report.valid_cell_name(params[:id])
+      @table = @report.valid_table_name(params[:table])
+      @clients = HudPathReport::Fy2020::DqClient.
```
Comment (eanders) [inline drivers/hud_path_report/app/controllers/hud_path_report/cells_controller.rb]: should this be `DqClient`?  Maybe PathClient?

File: drivers/hud_path_report/app/controllers/hud_path_report/cells_controller.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        joins(hud_reports_universe_members: { report_cell: :report_instance }).
+        merge(::HudReports::ReportCell.for_table(@table).for_cell(@cell)).
+        merge(::HudReports::ReportInstance.where(id: @report.id))
+      @name = "#{report_short_name} #{@question} #{@cell}"
```
Comment (eanders) [inline drivers/hud_path_report/app/controllers/hud_path_report/cells_controller.rb:19]: Do we need to sanitize this in any way?

File: drivers/hud_path_report/app/controllers/hud_path_report/base_controller.rb [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    private def path_for_questions(question)
+      hud_reports_path_questions_path(path_id: 0, question: question)
+    end
+
+    private def path_for_question_result(question, report: nil)
+      result_hud_reports_path_question_path(path_id: report&.id || 0, id: question)
+    end
+
+    private def path_for_report(report)
+      hud_reports_path_path(report)
```
Comment (eanders) [inline drivers/hud_path_report/app/controllers/hud_path_report/base_controller.rb:80]: gotta love a report called `path`  That could be tricky.

File: drivers/hud_path_report/app/controllers/hud_path_report/paths_controller.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        format.html do
+          @show_recent = params[:id].to_i.positive?
+          @questions = generator.questions.keys
+          @contents = @report&.completed_questions
+          @path_for_running = running_hud_reports_paths_path(link_params.except('action', 'controller'))
+        end
+        format.zip do
+          exporter = ::HudReports::ZipExporter.new(@report)
+          date = Date.current.strftime('%Y-%m-%d')
+          send_data exporter.export!, filename: "dq-#{date}.zip"
```
Comment (eanders) [inline drivers/hud_path_report/app/controllers/hud_path_report/paths_controller.rb]: maybe the filename should be `path-#{date}.zip`?

File: drivers/hud_path_report/app/controllers/hud_path_report/paths_controller.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    end
+
+    def new
+    end
+
+    def create
+      if @filter.valid?
+        @report = report_source.from_filter(@filter, report_name, build_for_questions: generator.questions.keys)
+        generator.new(@report).queue
+        redirect_to hud_reports_path_path(0)
```
Comment (eanders) [inline drivers/hud_path_report/app/controllers/hud_path_report/paths_controller.rb]: How hard is it to get rid of these 0s.  I feel like they've been biting us.  Shouldn't that just be a redirect to the index or to the new action?

File: drivers/hud_path_report/app/controllers/hud_path_report/cells_controller.rb (hunk 4 above)
Comment (awisema) [reply drivers/hud_path_report/app/controllers/hud_path_report/cells_controller.rb]: Good catch!

File: drivers/hud_path_report/app/controllers/hud_path_report/cells_controller.rb (hunk 5 above)
Comment (awisema) [reply drivers/hud_path_report/app/controllers/hud_path_report/cells_controller.rb:19]: `report short name` comes from us, and `@question` and `@cell` are sanitized above... So, I _think_ we are OK... However, I did find that `valid_table_name` was not going to accept the tables with dashes in their names...

File: drivers/hud_path_report/app/controllers/hud_path_report/paths_controller.rb (hunk 8 above)
Comment (awisema) [reply drivers/hud_path_report/app/controllers/hud_path_report/paths_controller.rb]: I created a new route to get rid of these -- the paths like `.../paths/0/questions/...` need some more thought...

Comment (eanders) [reply drivers/hud_path_report/app/controllers/hud_path_report/paths_controller.rb]: thank you, thank you, thank you!

File: drivers/hud_path_report/app/models/hud_path_report/generators/fy2020/base.rb [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        pending_associations = {}
+        batch.each do |client|
+          enrollment = last_enrollment(client)
+          next unless enrollment.present?
+
+          source_client = enrollment.client
+
+          max_disability_date = enrollment.disabilities.select { |d| d.InformationDate <= @report.end_date }.
+            map(&:InformationDate).max
+          disabilities_latest = enrollment.disabilities.select { |d| d.InformationDate == max_disability_date }
```
Comment (eanders) [inline drivers/hud_path_report/app/models/hud_path_report/generators/fy2020/base.rb:36]: Is there a reason not to use `max_by`? https://apidock.com/ruby/Enumerable/max_by

Comment (eanders) [reply drivers/hud_path_report/app/models/hud_path_report/generators/fy2020/base.rb:36]: Never mind, this is getting all records from the date.

**Review by eanders — CHANGES_REQUESTED** [review]
There seem to be some bugs with drilling down on the details for cells.

**Pull request merged**